### PR TITLE
Encoding performance improvements

### PR DESCRIPTION
**Pull request by tankyleo**
Make significant performance improvements to `BytesToHexIter` and `DisplayHex`, see individual commits for details on the exact performance improvements.

Also deletes `fn byte_to_hex_chars` as it repeats the functionality offered by `Table::byte_to_hex`.

### CODE REVIEW

**Comment by tankyleo** [comment]
This breaks #113, so I mark it as draft until #113 is merged, happy to be patient.

Please see https://github.com/tankyleo/hex-benchmarks for how I got the "Estimated Cycles" numbers in the commit messages.

Thank you for your time and feedback.

File: src/buf_encoder.rs
Comment (tcharding) [inline src/buf_encoder.rs:44]:
```suggestion
        let hex_bytes = case.table().byte_to_hex(byte);
        // SAFETY: Table::byte_to_hex returns only valid ASCII
        let hex_str = unsafe { core::str::from_utf8_unchecked(&hex_chars) };
        self.buf.push_str(hex_str);
```

File: src/iter.rs
Comment (tcharding) [inline src/iter.rs:228]: `char::from(low)` would be more clear here because makes it explicit that no data is being lost because it is infallible. As it is, the reader has to look up what type `low` is and then look up (or know) what casting that type to a `char` does.

Comment (tcharding) [reply src/iter.rs:228]: Some below for `high`.

**Review by tcharding — COMMENTED** [review]
Cheers, looks good. The second patch I'm ambivalent on, we typically try to avoid casts but in that case its super obvious what is going on. The `char` casting would be better done with `char::from` IMO.

File: src/buf_encoder.rs
Comment (tankyleo) [reply src/buf_encoder.rs:44]: `hex_chars` would be `hex_bytes` here, but overall I agree thank you !

**Comment by tankyleo** [comment]
> The second patch I'm ambivalent on, we typically try to avoid casts but in that case its super obvious what is going on. The char casting would be better done with char::from IMO.

I understand. Upon further thought, I'd be happy to drop the casts, and use `from` everywhere.

**Comment by tankyleo** [comment]
Rebase:
- Drop all casts, replace with `from`.
- Rename variables for clarity.
- Name the return type explicitly for clarity.

**Review by apoelstra — APPROVED** [review]
ACK b659ecaf3973c4e52870f64f1d7bcab65b4fab0c successfully ran local tests

**Review by tcharding — APPROVED** [review]
ACK b659ecaf3973c4e52870f64f1d7bcab65b4fab0c

**Comment by tcharding** [comment]
Nice, thanks. Did your github username change? I got confused seeing patches with a familiar name/email address for the author with a different username.

**Comment by tankyleo** [comment]
I mark it as ready in case you'd like to merge this before #113

**Comment by tcharding** [comment]
You can nominate your merge order preference if you want to.

**Comment by tankyleo** [comment]
@tcharding your call. I have a preference for #117 first. It seems to me kix might want to review #113 before it goes in, as he had some concerns in a prior PR.

**Comment by tcharding** [comment]
He has had two weeks already I believe. @apoelstra is the merger round here though so he will take care of the order. Cheers.

**Review by tcharding — APPROVED** [review]
ACK 29fb0a3bc53e0b39fc0a9f85fbb3f2662797b9fc

**Comment by apoelstra** [comment]
Nice. Will do this one.

cc @RCasatta you might be interested in trying this on your profiling.

**Review by apoelstra — APPROVED** [review]
ACK 29fb0a3bc53e0b39fc0a9f85fbb3f2662797b9fc successfully ran local tests

**Comment by RCasatta** [comment]
> cc @RCasatta you might be interested in trying this on your profiling.

In https://github.com/rust-bitcoin/hex-conservative/issues/115 I am interested in the opposite conversion, but thanks a lot for citing me, very nice perf improvement and TIL iai

**Pull request merged**
